src/Chapter4.hs

```diff
@@ -673,6 +722,26 @@ Specifically,
  ❃ Implement the function to convert Tree to list
 -}
 
+data BinaryTree a = EmptyNode | Node a (BinaryTree a) (BinaryTree a)
+  deriving (Show)
+
+insert :: Ord a => a -> BinaryTree a -> BinaryTree a
+insert x EmptyNode = Node x EmptyNode EmptyNode
+insert x (Node val l r) | x > val   = Node val l (insert x r)
+                        | otherwise = Node val (insert x l) r
+
+instance Functor BinaryTree where
+  fmap _ EmptyNode    = EmptyNode
+  fmap f (Node x l r) = Node (f x) (fmap f l) (fmap f r)
+
+reverseTree :: BinaryTree a -> BinaryTree a
+reverseTree EmptyNode    = EmptyNode
+reverseTree (Node x l r) = Node x (reverseTree r) (reverseTree l)
+
+treeToList :: BinaryTree a -> List a
+treeToList EmptyNode = Empty
+treeToList (Node x l r) =
+  append (append (treeToList l) (Cons x Empty)) (treeToList r)
```

Comment:
Is it possible to use fmap in any of these tree functions? I was trying to use it... but couldn't find a decent way to. Perhaps with a decent monad implementation we'd be able to use >>= to convert Tree to List? 🤔

Reply:
No, it's not possible to use `fmap` :disappointed: And the `Tree` data type is also not a `Monad`. This final boss combines tasks from different chapters: pattern-matching, custom data types, Functor from this chapter. If you want to play with functors, you can implement a function that increments each value in each tree node, but that's not so interesting :slightly_smiling_face: 

Here you can simplify the implementation a bit by using fewer appends and writing `append` in infix form:

```suggestion
  treeToList l `append` Cons x (treeToList r)
```